File: python/day11/day11.py
import re
import time
import heapq
from collections import deque, defaultdict, Counter
from functools import cache


input = open("../../inputs/day11.txt").read().splitlines()


def is_in_bounds(point, max_x, max_y):
    return (0 <= point[0] <= max_x) and (0 <= point[1] <= max_y)


def part1(lines):
    stones = [int(x) for x in lines[0].split()]
    for _ in range(25):
        new_stones = []
        for stone in stones:
            if stone == 0:
                new_stones.append(1)
            elif len(str(stone)) % 2 == 0:
                str_stone = str(stone)
                a = str_stone[: len(str_stone) // 2]
                b = str_stone[len(str_stone) // 2 :]
                new_stones.append(int(a))
                new_stones.append(int(b))
            else:
                new_stones.append(stone * 2024)
        stones = new_stones

    return len(stones)

@cache
def dfs(stone, days):
    if days == 0:
        return 1
    
    if stone == 0:
        return dfs(1, days - 1)
    elif len(str(stone)) % 2 == 0:
        str_stone = str(stone)
        a = str_stone[: len(str_stone) // 2]
        b = str_stone[len(str_stone) // 2 :]
        return dfs(int(a), days - 1) + dfs(int(b), days - 1)
    else:
        return dfs(stone * 2024, days - 1)

def part2(lines):
    stones = [int(x) for x in lines[0].split()]
    m = defaultdict(int)
    for stone in stones:
        m[stone] += 1
    for _ in range(75):
        new_m = defaultdict(int)
        for stone, num in m.items():
            if stone == 0:
                new_m[1] += num
            elif len(str(stone)) % 2 == 0:
                str_stone = str(stone)
                a = str_stone[: len(str_stone) // 2]
                b = str_stone[len(str_stone) // 2 :]
                new_m[int(a)] += num
                new_m[int(b)] += num
            else:
                new_m[stone * 2024] += num
        m = new_m

    print(sum(dfs(x, 75) for x in stones))
    return sum(m.values())


print("ans pt1:", part1(input))
print("***********part 2***************")
print(part2(input))

File: rust/src/bin/day11.rs
use std::cmp;
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let lines: Vec<String> = read_lines("../inputs/day11.txt")
        .unwrap()
        .map_while(Result::ok)
        .collect();
    part1(&lines);
    part2(&lines);
}
fn is_in_bounds(xi: i32, yi: i32, max_x: i32, max_y: i32) -> bool {
    (0 <= xi && xi <= max_x) && (0 <= yi && yi <= max_y)
}

fn part1(lines: &[String]) {
    println!("hasfd");
    let mut stones: Vec<i64> = lines
        .first()
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse::<i64>().unwrap())
        .collect();
    for _ in 0..25 {
        let mut new_stones = Vec::new();
        for stone in stones.iter() {
            if *stone == 0 {
                new_stones.push(1);
            } else if stone.to_string().len() % 2 == 0 {
                let str_stone = stone.to_string();
                let (a, b) = str_stone.split_at(str_stone.len() / 2);
                new_stones.push(a.parse().unwrap());
                new_stones.push(b.parse().unwrap());
            } else {
                new_stones.push(*stone * 2024);
            }
        }
        stones = new_stones;
    }
    println!("pt1: {}", stones.len());
}
fn part2(lines: &[String]) {
    let mut stones: Vec<i64> = lines
        .first()
        .unwrap()
        .split_whitespace()
        .map(|s| s.parse::<i64>().unwrap())
        .collect();
    let mut m: HashMap<i64, i64> = HashMap::new();
    for stone in stones {
        *m.entry(stone).or_insert(0) += 1;
    }
    let z = (0..75).fold(m, |acc, i| {
        let mut new_stones: HashMap<i64, i64> = HashMap::new();
        for (stone, num) in acc.iter() {
            if *stone == 0 {
                *new_stones.entry(1).or_insert(0) += num;
            } else if stone.to_string().len() % 2 == 0 {
                let str_stone = stone.to_string();
                let (a, b) = str_stone.split_at(str_stone.len() / 2);
                let x = a.parse().unwrap();
                let y = b.parse().unwrap();
                *new_stones.entry(x).or_insert(0) += num;
                *new_stones.entry(y).or_insert(0) += num;
            } else {
                let y = *stone * 2024;
                *new_stones.entry(y).or_insert(0) += num;
            }
        }
        new_stones
    });
    println!("pt2: {}", z.values().sum::<i64>());
}
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
